### ISSUE TITLE
Rolling update puts nodes into "not ready"

### ISSUE THREAD

**Issue by recollir** [kubernetes/kops#4946]
1.  Kops version 1.8.0
2. Kubernetes version 1.8.6
3. AWS (3 masters and 3 nodes)

4. kops edit followed by kops update and kops rolling-update. kops edit to add configuration flags for the apiserver (dex related). Also tried kops rolling-update --instance-group <master...> to only update one master at a time.

5. Nodes become "not ready" in an unpredictable way. Sometimes no node is affected. Sometimes one node becomes "not ready" and recovers after a few minutes. Sometimes all nodes are "not ready" for a longer period. Up to 15 minutes. While the masters report ready. During this time the workload on the cluster is not accessible.

6. Nothing: a non-breaking rolling update without affecting nodes or the workload.

7. Starting config: https://gist.github.com/recollir/9e9b4b0b426ef77014083f1839c123d6
   Added via kops edit before the rolliing-update: https://gist.github.com/recollir/da9fd8a123b58f555f2e4321093e9d46

8. https://gist.github.com/recollir/5b19d543adaa50b1889aabafeb77b847 

9. A couple of times I observed that after the rolling update the ELB for the API server was missing AZ attached to it.

**Comment by johanneswuerbach** [comment]
In our case manually restarting the kubelet helped, do you have the logs of an affected node?

**Comment by recollir** [comment]
Not from the current test runs (cluster has been deleted and created a couple of times). But I can recreate this, keep the logs and attach them to this issue.

Nevertheless, I would be interested how to prevent this from happening. We need to do some changes to a production cluster where a restart of a kubelet seems rather inappropriate.

**Comment by johanneswuerbach** [comment]
Understandably, we are also started this this upgrading to 1.8 and I’m currently debugging what could cause this.

It looks in our case the kubelet tries to connect to an old API server IP so either its caching an old record or the record wasn’t updated correctly.

**Comment by recollir** [comment]
Thanks for the pointing this out. I will see if this is the same for us first thing in the morning. My TZ is CEST.

**Comment by justinsb** [comment]
Thanks for reporting & sorry about the problem.

Was this with a gossip DNS (.k8s.local) or a "real" Route53 DNS name?

**Comment by johanneswuerbach** [comment]
In our case a real Route53 record.

**Comment by recollir** [comment]
Real Route53 DNS name.

**Comment by recollir** [comment]
@johanneswuerbach seems the same for us - kubelet trying to connect to an old API server IP. Trying to verify this now.

**Comment by johanneswuerbach** [comment]
We also hit this on another node again:
```
kubelet[1270]: E0410 13:01:58.506984    1270 kubelet_node_status.go:390] Error updating node status, will retry: error getting node "ip-xxx.internal": Get https://api.internal.xxx/api/v1/nodes/ip-xxx.ec2.internal?resourceVersion=0: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
kubelet[1270]: E0410 13:02:08.507374    1270 kubelet_node_status.go:390]1270 kubelet_node_status.go:390] Error updating node status, will retry: error getting node "ip-xxx.ec2.internal": Get https://api.internal.xxx/api/v1/nodes/ip-xxx.ec2.internal: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
and eventually
kubelet[1270]: E0410 13:02:38.508388    1270 kubelet_node_status.go:382] Unable to update node status: update node status exceeds retry count
```

**Comment by recollir** [comment]
That is exactly the error msg I see. It starts to appear when the old IP address is removed from the A record for api.internal.xxx and the new IP address for the new master is added. Sometimes after the first master, sometimes after the second master.

**Comment by lkysow** [comment]
It's probably due to: https://github.com/kubernetes/kubernetes/issues/41916#issuecomment-312428731 where the kubelet caches the IP of the old master nodes. That's why a restart fixes it.

**Comment by sstarcher** [comment]
I had the same issue today with 1.9.0-beta-2

**Comment by lkysow** [comment]
I think the best practice is to set up an internal ELB that fronts the masters and have the API url point to that, the same way it's done for the external api. Is that possible with kops right now?

**Comment by recollir** [comment]
I don’t think so. The type load balancer for the api in the spec refers to the client (kubectl) AFAIK. At least in a quick test I still got DNS round robin based entries for the api that the kubelet used.

I think as well an ELB for the kubelet to connect to would be the “right” way to go. At least it is the way kubeadm does HA nowadays (if though manually still). The ELB would detect that a master is gone through the health check, break the connection and force the kubelet to reconnect, wouldn’t it?

What would be needed? How much work would it be? Are there any pointers to start? I wouldn’t mind to give it a try. But would need instructions.

**Comment by sstarcher** [comment]
Just did a rolling update from 1.9.0-beta-2 to 1.9.0 and the same issue all of my nodes go from Ready to not ready.

@chrislovecnm have you tried a master rolling update with 1.9.0 by chance on AWS?

**Comment by sstarcher** [comment]
I can confirm it happens for 15mins or until the kubelet is restarted.

**Comment by lkysow** [comment]
15m is what is expected for the issue with kubelet caching IPs: https://github.com/kubernetes/kubernetes/issues/41916#issuecomment-312428731

**Comment by recollir** [comment]
Just realised that the same problem affects the kube-proxy, btw.

**Comment by jaredallard** [comment]
Ran into this today with a beta environment deploy -- thankfully nothing broke in our production env, but certainly not a good sign... Any detailed fixes for this? `kubelet` restart sure, but on each node?

**Comment by recollir** [comment]
Just to add another occasion where this can happen: when "updating" from kops 1.8 to kops 1.9 and performing the required rolling-update. As first all masters are restarted/recreated, the nodes can become not ready if kubelet/kube-proxy was/is talking to the corresponding, restarting master.

**Comment by zachaller** [comment]
We are also being hit by this and its causing our own api's to have downtime when the masters come back up. I do think that putting an elb on the internal api endpoint would help in this case as well.

**Comment by mattatcha** [comment]
Looks like a fix has been merged and a pr is open to backport to 1.9 
fix: https://github.com/kubernetes/kubernetes/pull/63492
1.9 backport: https://github.com/kubernetes/kubernetes/pull/63832

**Comment by recollir** [comment]
Approved and cherry-picked as well.
